lib/src/misc/frame_fit.dart

```diff
@@ -0,0 +1,207 @@
+  final bool forceIntegerZoomLevel;
+
+  const FitBounds({
+    required this.bounds,
+    this.padding = EdgeInsets.zero,
+    this.maxZoom = 17.0,
+    this.inside = false,
+    this.forceIntegerZoomLevel = false,
+  });
+
+  @override
+  FlutterMapState fit(FlutterMapState mapState) {
+    final paddingTL = CustomPoint<double>(padding.left, padding.top);
+    final paddingBR = CustomPoint<double>(padding.right, padding.bottom);
+
+    final paddingTotalXY = paddingTL + paddingBR;
+
+    var newZoom = getBoundsZoom(mapState, paddingTotalXY);
+    newZoom = math.min(maxZoom, newZoom);
+
+    final paddingOffset = (paddingBR - paddingTL) / 2;
+    final swPoint = mapState.project(bounds.southWest, newZoom);
+    final nePoint = mapState.project(bounds.northEast, newZoom);
+
+    final CustomPoint<double> projectedCenter;
+    if (mapState.rotation != 0.0) {
+      final swPointRotated = swPoint.rotate(-mapState.rotationRad);
+      final nePointRotated = nePoint.rotate(-mapState.rotationRad);
+      final centerRotated =
+          (swPointRotated + nePointRotated) / 2 + paddingOffset;
+
+      projectedCenter = centerRotated.rotate(mapState.rotationRad);
+    } else {
+      projectedCenter = (swPoint + nePoint) / 2 + paddingOffset;
+    }
+
+    final center = mapState.unproject(projectedCenter, newZoom);
+    return mapState.withPosition(
+      center: center,
+      zoom: newZoom,
+    );
+  }
+
+  double getBoundsZoom(
+    FlutterMapState mapState,
+    CustomPoint<double> pixelPadding,
+  ) {
+    final min = mapState.minZoom ?? 0.0;
+    final max = mapState.maxZoom ?? double.infinity;
+    final nw = bounds.northWest;
+    final se = bounds.southEast;
+    var size = mapState.nonRotatedSize - pixelPadding;
+    // Prevent negative size which results in NaN zoom value later on in the calculation
+    size = CustomPoint(math.max(0, size.x), math.max(0, size.y));
+    var boundsSize = Bounds(
+      mapState.project(se, mapState.zoom),
+      mapState.project(nw, mapState.zoom),
+    ).size;
+    if (mapState.rotation != 0.0) {
+      final cosAngle = math.cos(mapState.rotationRad).abs();
+      final sinAngle = math.sin(mapState.rotationRad).abs();
+      boundsSize = CustomPoint<double>(
+        (boundsSize.x * cosAngle) + (boundsSize.y * sinAngle),
+        (boundsSize.y * cosAngle) + (boundsSize.x * sinAngle),
+      );
+    }
+
+    final scaleX = size.x / boundsSize.x;
+    final scaleY = size.y / boundsSize.y;
+    final scale = inside ? math.max(scaleX, scaleY) : math.min(scaleX, scaleY);
+
+    var boundsZoom = mapState.getScaleZoom(scale, mapState.zoom);
+
+    if (forceIntegerZoomLevel) {
+      boundsZoom =
+          inside ? boundsZoom.ceilToDouble() : boundsZoom.floorToDouble();
+    }
+
+    return math.max(min, math.min(max, boundsZoom));
+  }
+}
+
+class FitCoordinates extends FrameFit {
+  final List<LatLng> coordinates;
+  final EdgeInsets padding;
+  final double maxZoom;
+  final bool inside;
```

Comment:
I included this `inside` parameter in my `fitCoordinates` PR for parity with `fitBounds`, but upon further reflection I think that may have been a mistake on my part.

For `fitBounds`, so far as I can tell `inside: true` basically means "fit the map frame inside the bounds instead of fitting the bounds inside the map frame" (aside: it's possible my fix for rotation in https://github.com/fleaflet/flutter_map/pull/1550 didn't properly handle `inside: true`, so I will look into that further and maybe open a separate issue to discuss that).

For `fitCoordinates`, I migrated that implementation fairly directly but I now suspect the result is actually incoherent; as discussed above, a list of coordinates does not define a "bounds" in any unambiguous sense, so the idea of fitting the map frame _inside_ the coordinates doesn't really make sense. So I think I would advocate removing this parameter from `FitCoordinates`.

My apologies for causing confusion by including it in the first place!

Reply:
No problem! I've remove `inside` from FitCoordinates. If you figure out the rotation fix for FitBounds I'm happy to update it in this PR.

Reply:
Continued at https://github.com/fleaflet/flutter_map/pull/1551#issuecomment-1597631016 and after...